Updates for CMSSW_9_4_X

pieterdavid opened the pull request:
I checked out the Framework in CMSSW_9_4_9, removed everything that said 80X or 'Moriond17' etc., and tried to make things run... so there will be differences, and I may have been too enthusiastically removing things in some places - my main aim was to provide a "clean" starting point, but feel free to point out if something doesn't make sense.

Anticipating a few questions:
- on-demand scheduling requires a `cms.Task` to be associated with the process or path for each `edm::EDProducer`. This required quite a bit of trial-and-error with redoJEC etc., but the tests should run now
- I turned off EGM regression and smearing for now, because the way this works changed (new recipes at https://twiki.cern.ch/twiki/bin/view/CMS/EgammaMiniAODV2 and https://twiki.cern.ch/twiki/bin/view/CMS/EgammaPostRecoRecipes) and I haven't implemented the new recipes yet.

pieterdavid committed:
Remove merged recipes and move to the latest release

pieterdavid committed:
Add description of branches to README

pieterdavid committed:
Move correction versions out of Framework.py

pieterdavid committed:
Remove 2016-specific code from producers and config

pieterdavid committed:
KaMuCa is a build dependency

pieterdavid committed:
gcc 6.3 fix: include string

pieterdavid committed:
Use cms.Task for on-demand producers

pieterdavid committed:
Exclude probcc deepcsv category (for 9_3_X)

pieterdavid committed:
Turn off electron regression and smearing (waiting for recipes)

pieterdavid committed:
More CHS inputs for MET calculation

pieterdavid committed:
put back electron ID (working points and MVA classifiers)

pieterdavid committed:
use slc6_amd64_gcc630 SCRAM_ARCH (for jenkins)

pieterdavid committed:
Turn off electron regression and smearing everywhere

pieterdavid committed:
Update README, arch and release for 94X

OlivierBondu commented on README.md:
```diff
@@ -18,12 +23,14 @@ Common framework for all cp3-llbb analyses
-wget https://raw.githubusercontent.com/cp3-llbb/Framework/CMSSW_8_0_6p/setup_project_with_framework.sh
-source setup_project_with_framework.sh
+wget https://raw.githubusercontent.com/cp3-llbb/Framework/CMSSW_9_4_X/setup_project_with_framework.sh
+source setup_project_with_framework.sh --branch BRANCH ## select a branch
```
maybe we could assume here the default branch should be 94X ? I would expect newcomers (and myself) to copy paste blindly in a terminal without checking that there is need to replace BRANCH by whatever

pieterdavid replied:
Indeed, that makes sense

OlivierBondu commented on README.md:
```diff
@@ -18,12 +23,14 @@ Common framework for all cp3-llbb analyses
-Through the options `--branch NAME` and `--pr ID`, a project area for a different version can also be created.
+[This script](setup_project_with_framework.sh) will set up a CMSSW release area,
+apply the recipes in [``bootstrap_jenkins.sh``](bootstrap_jenkins.sh) and [``jenkins_postbuild.sh``](jenkins_postbuild.sh),
+perform an initial build, and add your and your colleagues' forks on GitHub as remotes for your ``Framework`` clone.
```
speaking of which, maybe we could profit from this cleanup PR to cleanup [these](https://github.com/cp3-llbb/Framework/blob/CMSSW_8_0_6p/first_setup.sh#L17-L27) ? Do you know if there would exist an api to get the list automatically ? This would remove the pain of updating the list...

pieterdavid replied:
Of course there is... it return json, so I made the script in python. It takes "active" forks as all those that have been pushed to during the last year, and should also work for other packages - assuming the base repository is set as 'upstream', or under 'cp3-llbb'. There maybe other things missing.

OlivierBondu commented on interface/ElectronsProducer.h:
```diff
@@ -52,8 +52,6 @@ class ElectronsProducer: public LeptonsProducer<pat::Electron>, public Identifia
         BRANCH(trackIso, std::vector<float>);
 
-        BRANCH(expectedMissingInnerHits, std::vector<unsigned int>);
-
```
I assume you are removing this because it will be in the ID itself ?

pieterdavid replied:
yes, I suppose that with the reprocessing there will be one ID for 2016-2017-2018 that can be used without additional ad-hoc cuts.

OlivierBondu commented on python/ElectronsProducer.py:
```diff
@@ -9,34 +9,10 @@
-                'egmGsfElectronIDs:cutBasedElectronHLTPreselection-Summer16-V1',
-                'egmGsfElectronIDs:mvaEleID-Spring16-GeneralPurpose-V1-wp90',
-                'egmGsfElectronIDs:mvaEleID-Spring16-GeneralPurpose-V1-wp80',
-                'egmGsfElectronIDs:mvaEleID-Spring16-HZZ-V1-wpLoose'
```
there is no ID at all ??

pieterdavid replied:
There are 'Fall17' 94X IDs to be used for 2017 and 2018, but the twiki doesn't mention reprocessed 2016 data.

OlivierBondu commented on bootstrap_jenkins.sh:
```diff
@@ -22,15 +22,13 @@ function cms-merge-commit()
 
 echo "---> Adding KaMuCa"
-git clone -o upstream https://github.com/bachtis/analysis.git -b KaMuCa_V4 KaMuCa 
+git clone -o upstream https://github.com/bachtis/analysis.git -b KaMuCa_V4 KaMuCa
```
KaMuCa has not been updated in forever, it most probably won't be up to date either for 94X ?

pieterdavid replied:
I only checked it was the latest... disabled now

pieterdavid commented:
I went through the diffs...
For `unit_tests_data.py` the differences are minor: all scale factors gone (I removed them), `electron_expectedMissingInnerHits`, `(fat)jet_pfDeepCSVJetTags_probcc` (it seems not to be there in this version of the deepcsv) and `muon_isMedium2016`. The former and latter will probably be needed still, so I can put them back easily. There are also differences in `jet_pfDeepCSVJetTags_probc` (maybe the merged `probc` and `probcc`?), and in the electron IDs because I removed them... maybe in hindsight I shouldn't have.
For `unit_test_mc_with_db.py` the same, plus the PU weights which I also removed.
For `unit_test_mc.py` there are more differences... I turned off the electron smearing and regression (the recipes changed, see [this twiki](https://twiki.cern.ch/twiki/bin/viewauth/CMS/EgammaMiniAODV2#Energy_Scale_and_Smearing), I didn't get to implementing that).
Also the jet, jet systematic variation and MET branches changed. For the former two I suppose that would be because we pick up newer calibrations/recipes somehow, for the MET I'm reasonably certain (I saw code changes that needed the config to be adjusted). Does that make sense?

OlivierBondu commented on python/Framework.py:
```diff
@@ -69,7 +67,6 @@ def __init__(self, options, verbose=True):
 
         process.options = cms.untracked.PSet(
                 wantSummary = cms.untracked.bool(True),
-                allowUnscheduled = cms.untracked.bool(True)
```
why not keep the option explicit and put it to false instead ?

pieterdavid replied:
see https://twiki.cern.ch/twiki/bin/view/CMSPublic/SWGuideUnscheduledExecution#Usage (it's always true, so this does not have any effect)

OlivierBondu commented on python/Tools.py:
```diff
@@ -185,7 +195,22 @@ def recorrect_met(process, isData, metCollection, jetCollection):
+            , [ "patPFMetT1Txy", "patPFMetTxyCorr" ]
+        ))+
+        (["ak4PFCHS{0}Corrector".format(corrType) for corrType in ("L1FastL2L3", "L1Fastjet", "L2Relative", "L3Absolute", "L1FastL2L3Residual", "Residual")] if isData else [])+
+        [ "patCaloMet", "patCHSMet", "pfMetCHS", "pfCHS", "patTrkMet", "pfMetTrk", "pfTrk", "metrawCaloNewJEC" ])
```
I will trust you on this one, I don't know JEC nor MET :smile:

pieterdavid replied:
This is pure trial&error (but unfortunately needed with tasks, until we get a cms.Task-proof version of it): I went up to the point where I don't get errors about products that don't exist, and neither about modules that are not defined

OlivierBondu commented on test/unit_tests_mc.py:
```diff
@@ -11,9 +11,9 @@
-framework.applyMuonCorrection("kamuca")
-framework.applyElectronRegression()
-framework.applyElectronSmearing()
+framework.applyMuonCorrection("kamuca", tag="MC_80X_13TeV")
```
please comment this line out

OlivierBondu reviewed (changes requested):
Thanks!

sorry for the late review, only minor changes :)

pieterdavid committed:
default to 94X branch in README for 94X

pieterdavid committed:
update forks list in first_setup

pieterdavid committed:
replace first_setup.sh with scripts/updateremotes, which uses the github API to search for forks

pieterdavid committed:
Add Fall17 electron IDs

pieterdavid committed:
Disable KaMuCa

OlivierBondu commented:
the `updateremotes` script is just perfect :) thanks for the other fixes !

Do you know why Jenkins fails ? Is it just the outdated ref trees ?

pieterdavid commented:
I changed the definition of the target branch, so reopening as a new PR, https://github.com/cp3-llbb/Framework/pull/284
